tasm-lib/src/mmr/authentication_struct/root_from_authentication_struct.rs

```diff
@@ -0,0 +1,1040 @@
+use triton_vm::prelude::*;
+use twenty_first::math::x_field_element::EXTENSION_DEGREE;
+
+use crate::data_type::DataType;
+use crate::library::Library;
+use crate::mmr::authentication_struct::derive_challenges::DeriveChallenges;
+use crate::prelude::BasicSnippet;
+
+pub struct RootFromAuthenticationStruct;
+
+impl RootFromAuthenticationStruct {
+    fn indexed_leaf_element_type() -> DataType {
+        DataType::Tuple(vec![DataType::U64, DataType::Digest])
+    }
+}
+
+impl BasicSnippet for RootFromAuthenticationStruct {
+    fn inputs(&self) -> Vec<(DataType, String)> {
+        vec![
+            (DataType::U32, "tree_height".to_owned()),
+            (
+                DataType::List(Box::new(DataType::Digest)),
+                "auth_struct".to_owned(),
+            ),
+            (
+                DataType::List(Box::new(Self::indexed_leaf_element_type())),
+                "indexed_leafs".to_owned(),
+            ),
+        ]
+    }
+
+    fn outputs(&self) -> Vec<(DataType, String)> {
+        vec![(DataType::Digest, "root".to_owned())]
+    }
+
+    fn entrypoint(&self) -> String {
+        "tasmlib_mmr_root_from_authentication_struct".to_owned()
+    }
+
+    fn code(&self, library: &mut Library) -> Vec<LabelledInstruction> {
+        let alpha_challenge_pointer = library.kmalloc(EXTENSION_DEGREE as u32);
+        let alpha_challenge_pointer_write = alpha_challenge_pointer.write_address();
```

Comment:
nit:
- The return value of `kmalloc` is not a pointer (anymore), but a static allocation.
- Many of the `*_{read, write}` variables are used exactly once, meaning their declaration could be omitted, resulting in less clutter and better readability of the assembly, in my opinion.

For example, the allocations could be used like:

```rust
triton_asm!{
  push {alpha_challenge_alloc.write_address()}
  write_mem {alpha_challenge_alloc.num_words()} // no repetition of allocation size
  pop 1
}
```

Reply:
That's a cool idea. It's probably what the `StaticAllocation` methods are used for almost exclusively anyway. That said, I don't feel strongly about the initial comment of this thread. 😌